multi/mcxn94x-multi/cpu.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * Phoenix-RTOS
+ *
+ * MCX N94x M33 CPU core driver
+ *
+ * Copyright 2025 Phoenix Systems
+ * Author: Daniel Sawka
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/file.h>
+#include <sys/interrupt.h>
+#include <sys/minmax.h>
+#include <sys/msg.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+
+#include <libtty-lf-fifo.h>
+
+#include "common.h"
+#include "dev.h"
+
+#define REG_IRQ_OFFSET 4
+/* clang-format off */
+enum {
+	reg_irq = 0, reg_irq_set, reg_irq_clr, mutex = 62
+};
+/* clang-format on */
+
+#define REG_IRQ(CPUID, REG) (*(common.mailbox + (REG_IRQ_OFFSET * (CPUID)) + (REG)))
+
+
+static struct {
+	uint8_t rx_buf[64];
+	lf_fifo_t rx_fifo;
+	handle_t rx_lock;
+	handle_t tx_lock;
+	handle_t irq_lock;
+	handle_t rx_cond;
+
+	volatile uint32_t *mailbox;
+	unsigned int major;
+	uint8_t coreid_self;
+	uint8_t coreid_remote;
+} common;
+
+
+static int mailbox_irqHandler(unsigned int n, void *arg)
+{
+	size_t consumed = 0;
+	uint32_t value = REG_IRQ(common.coreid_self, reg_irq);
+
+	if (value != 0) {
+		REG_IRQ(common.coreid_self, reg_irq_clr) = value;
+
+		size_t rx_len = (value >> 24) & 0x3;
+
+		for (; consumed < rx_len; consumed++) {
+			uint8_t byte = (value >> (8 * consumed)) & 0xff;
+
+			/* TODO: Data is overwritten if rx fifo is full. Consider disabling
+			the interrupt somehow and leaving the value in the register until
+			the application reads from rx fifo. Also, fifo needs to be more
+			generic and at least have function to get its current length. */
+			lf_fifo_push(&common.rx_fifo, byte);
+		}
+	}
+
+	return (consumed > 0) ? 1 : -1;
+}
+
+
+static ssize_t fifoRead(unsigned char *buf, size_t size, unsigned int blocking)
+{
+	size_t bytes_read = 0;
+
+	if (buf == NULL || size == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple readers */
+	mutexLock(common.rx_lock);
+
+	if (blocking != 0) {
+		while (bytes_read < size) {
+			mutexLock(common.irq_lock);
+			while (lf_fifo_empty(&common.rx_fifo) != 0) {
+				condWait(common.rx_cond, common.irq_lock, 0);
```

Comment:
I don't think we need separate blocks for blocking/non-blocking
```suggestion
				if (!blocking) {
					mutexLock(common.irq_lock);
					return bytes_read ? bytes_read : -EGAIN;
				}
				condWait(common.rx_cond, common.irq_lock, 0);
```

Reply:
I wanted to avoid locking `irqLock` unnecessarily in non-blocking call. Your suggestion makes code cleaner, to opted for it